File: BST/Swift/BST.swift
//
// Created by Shahriar Nasim Nafi on 13/1/22.
//

class BST<T: Comparable> {
    private var value: T
    private var left: BST?
    private var right: BST?

    init(value: T) {
        self.value = value
    }

    /**
     Insert new value to the `BST`
     __Average__: O(logn) time | O(1) space
     __Worst__: O(n) time | O(1) space
     */
    /// - Parameter value: value to add
    /// - Returns: Current `BST`
    func insert(_ value: T) -> BST {
        var current: BST? = self
        while current != nil {
            if current!.value > value {
                if (current!.left != nil) {
                    current = current?.left
                } else {
                    current!.left = BST(value: value)
                    break
                }
            } else {
                if (current!.right != nil) {
                    current = current?.right
                } else {
                    current!.right = BST(value: value)
                    break
                }
            }
        }

        return self
    }

    /// Check if value contains in the `BST`
    ///  __Average__: O(logn) time | O(1) space
    ///  __Worst__: O(n) time | O(1) space
    /// - Parameter value: value  to check
    /// - Returns: `true` if exists else `false`
    func contains(_ value: T) -> Bool {
        var current: BST? = self
        while current != nil {
            if (current!.value > value) {
                current = current?.left
            } else if (current!.value < value) {
                current = current?.right
            } else {
                return true
            }
        }

        return false
    }

    /// Remove specific value
    ///  __Average__: O(logn) time | O(1) space
    ///  __Worst__: O(n) time | O(1) space
    /// - Parameters:
    ///   - value: value want to remove
    ///   - parent: from the node want to remove else from current node
    /// - Returns: Current `BST`
    func remove(_ value: T, parent: BST? = nil) -> BST {
        var current: BST? = self
        var _parent: BST? = parent
        while current != nil {
            if (current!.value > value) {
                _parent = current
                current = current?.left
            } else if (current!.value < value) {
                _parent = current
                current = current?.right
            } else {
                if current!.left != nil && current!.right != nil {
                    current?.value = current!.right!.getMinValue()
                    current!.right!.remove(current!.value, parent: current)
                } else if (_parent == nil) {
                    if current!.left != nil {
                        current?.value = current!.left!.value
                        current?.right = current?.left?.right
                        current?.left = current?.left?.left
                    } else if current!.right != nil {
                        current?.value = current!.right!.value
                        current?.left = current?.right?.left
                        current?.right = current?.right?.right
                    }
                } else if _parent!.left === current! {
                    _parent?.left = current?.left != nil ? current?.left : current?.right
                } else if _parent!.right === current {
                    _parent?.right = current?.right != nil ? current?.right : current?.left
                }
                break
            }
        }
        return self
    }

    /// Remove specific value
    ///  __Average__: O(logn) time | O(1) space
    ///  __Worst__: O(n) time | O(1) space
    /// - Parameters:
    ///   - value: value want to remove
    ///   - parent: from the node want to remove else from current node
    /// - Returns: `true` if removed else `false`
    func removed(_ value: T, parent: BST? = nil) -> Bool {
        var current: BST? = self
        var _parent: BST? = parent
        while current != nil {
            if (current!.value > value) {
                _parent = current
                current = current?.left
            } else if (current!.value < value) {
                _parent = current
                current = current?.right
            } else {
                if current!.left != nil && current!.right != nil {
                    current?.value = current!.right!.getMinValue()
                    current!.right!.remove(current!.value, parent: current)
                } else if (_parent == nil) {
                    if current!.left != nil {
                        current?.value = current!.left!.value
                        current?.right = current?.left?.right
                        current?.left = current?.left?.left
                    } else if current!.right != nil {
                        current?.value = current!.right!.value
                        current?.left = current?.right?.left
                        current?.right = current?.right?.right
                    }
                } else if _parent!.left === current! {
                    _parent?.left = current?.left != nil ? current?.left : current?.right
                } else if _parent!.right === current {
                    _parent?.right = current?.right != nil ? current?.right : current?.left
                }
                return true
            }
        }
        return false
    }

    /// Get the minimum value from the node
    ///  __Average__: O(logn) time | O(1) space
    ///  __Worst__: O(n) time | O(1) space
    /// - Returns: value
    func getMinValue() -> T {
        var current: BST? = self
        while current?.left != nil {
            current = current?.left
        }
        return current!.value
    }
}

File: BST/CPP/BST.cpp
//
// Created by Shahriar Nasim Nafi on 13/1/22.
//

#include <iostream>

using namespace std;

template<class T>
class BST {
    T value;
    BST *left = NULL;
    BST *right = NULL;

public:
    explicit BST(T value) {
        this->value = value;
    }

    // Avareage: O(logn) time | O(1) space
    // Worst: O(n) time | O(1) space
    // Insert new value & return BST. So that we can use bst->insert(3)->insert(2)
    BST *insert(T value) {
        BST *current = this;
        while (true) {
            if (value < current->value) {
                if (current->left) {
                    current = current->left;
                } else {
                    current->left = new BST(value);
                    break;
                }
            } else {
                if (current->right) {
                    current = current->right;
                } else {
                    current->right = new BST(value);
                    break;
                }
            }
        }
        return this;
    }

    // Avareage: O(logn) time | O(1) space
    // Worst: O(n) time | O(1) space
    // check if contains
    bool contain(T value) {
        BST *current = this;
        while (current) {
            if (value < current->value) {
                current = current->left;
            } else if (value > current->value) {
                current = current->right;
            } else {
                return true;
            }
        }
        return false;
    }

    // Avareage: O(logn) time | O(1) space
    // Worst: O(n) time | O(1) space
    // Remove specific value & return BST. So that we can use bst->remove(3)->remove(2)
    BST *remove(T value, BST *parent = NULL) {
        BST *current = this;
        while (current) {
            if (value < current->value) {
                parent = current;
                current = current->left;
            } else if (value > current->value) {
                parent = current;
                current = current->right;
            } else {
                if (current->left && current->right) {
                    current->value = current->right->getMinValue();
                    current->right->remove(current->value, current);
                } else if (!parent) {
                    if (current->left) {
                        current->value = current->left->value;
                        current->right = current->left->right;
                        current->left = current->left->left;
                    } else if (current->right) {
                        current->value = current->right->value;
                        current->left = current->right->left;
                        current->right = current->right->right;
                    } else {
                        current->value = NULL;
                    }
                } else if (parent->left == current) {
                    parent->left = current->left ? current->left : current->right;
                } else if (parent->right == current) {
                    parent->right = current->right ? current->right : current->left;
                }
                break;
            }
        }
        return this;
    }

    // Avareage: O(logn) time | O(1) space
    // Worst: O(n) time | O(1) space
    // Remove specific value & return if it is removed or false if not is found
    bool removed(T value, BST *parent = NULL) {
        BST *current = this;
        while (current) {
            if (value < current->value) {
                parent = current;
                current = current->left;
            } else if (value > current->value) {
                parent = current;
                current = current->right;
            } else {
                if (current->left && current->right) {
                    current->value = current->right->getMinValue();
                    current->right->remove(current->value, current);
                } else if (!parent) {
                    if (current->left) {
                        current->value = current->left->value;
                        current->right = current->left->right;
                        current->left = current->left->left;
                    } else if (current->right) {
                        current->value = current->right->value;
                        current->left = current->right->left;
                        current->right = current->right->right;
                    } else {
                        current->value = NULL;
                    }
                } else if (parent->left == current) {
                    parent->left = current->left ? current->left : current->right;
                } else if (parent->right == current) {
                    parent->right = current->right ? current->right : current->left;
                }
                return true;
            }
        }
        return false;
    }

    // Avareage: O(logn) time | O(1) space
    // Worst: O(n) time | O(1) space
    // Get the minimum value of a node
    T getMinValue() {
        BST *current = this;
        while (current->left) {
            current = current->left;
        }
        return current->value;
    }

    void print(BST *bst = NULL, int level = 1) {
        if (bst == NULL)
            return;
        for (int i = 0; i < level; i++) {
            printf(i == level - 1 ? "|-" : "  ");
        }
        printf("%d\n", bst->value);
        print(bst->left, level + 1);
        print(bst->right, level + 1);
    }

};

